usr/man/man/usr/share/man/man4/nsswitch.conf.4
'\" te
.\" Copyright (c) 2008, 2011, Oracle and/or its affiliates. All rights reserved.
.TH nsswitch.conf 4 "16 Aug 2011" "SunOS 5.11" "File Formats"
.SH NAME
nsswitch.conf \- configuration file for the name service switch
.SH SYNOPSIS
.LP
.nf
\fBsvc:/system/name-service/switch\fR
.fi

.LP
.nf
\fB/etc/nsswitch.conf\fR
.fi

.SH DESCRIPTION
.sp
.LP
The operating system uses a number of databases of information about hosts, ipnodes, users (\fBpasswd\fR(4), \fBshadow\fR(4), and \fBuser_attr\fR(4)), and groups. Data for these can come from a variety of \fBsources: hostnames\fR and host addresses, for example, can be found in \fB/etc/hosts\fR, \fBNIS\fR, \fBLDAP\fR, \fBDNS\fR, or Multicast \fBDNS\fR. Zero or more sources can be used for each database; the sources and their lookup order are specified in svc:/system/name-service/switch service. For the purposes of backwards compatibility, the \fB/etc/nsswitch.conf file\fR is regenerated from the SMF properties configured in the \fBsvc:/system/name-service/switch\fR service. The \fB/etc/nsswitch.conf\fR file is considered obsolete.
.sp
.LP
The following databases use the \fBswitch\fR file:
.sp

.sp
.TS
tab();
cw(2.18i) cw(3.32i) 
lw(2.18i) lw(3.32i) 
.
DatabaseUsed By
\fBalias\fR\fBsendmail\fR(1M)
\fBauth_attr\fR\fBgetauthnam\fR(3C)
\fBautomount\fR\fBautomount\fR(1M)
\fBbootparam\fR\fBrpc.bootparamd\fR(1M)
\fBether\fR\fBethers\fR(3SOCKET)
\fBgroup\fR\fBgetgrnam\fR(3C)
\fBhost\fRT{
\fBgethostbyname\fR(3NSL), \fBgetaddrinfo\fR(3SOCKET). See \fBInteraction with netconfig\fR.
T}
\fBnetgroup\fR\fBinnetgr\fR(3C)
\fBnetmask\fR\fBifconfig\fR(1M)
\fBnetwork\fR\fBgetnetbyname\fR(3SOCKET)
\fBpasswd\fRT{
\fBgetpwnam\fR(3C), \fBgetspnam\fR(3C), \fBgetusernam\fR(3C)
T}
\fBprof_attr\fR\fBgetprofnam\fR(3C), \fBgetexecprof\fR(3C)
\fBproject\fRT{
\fBgetprojent\fR(3PROJECT), \fBgetdefaultproj\fR(3PROJECT), \fBinproj\fR(3PROJECT), \fBnewtask\fR(1), \fBsetproject\fR(3PROJECT)
T}
\fBprotocol\fR\fBgetprotobyname\fR(3SOCKET)
\fBpublickey\fR\fBgetpublickey\fR(3NSL), \fBsecure_rpc\fR(3NSL)
\fBrpc\fR\fBgetrpcbyname\fR(3NSL)
\fBservice\fR\fBgetservbyname\fR(3SOCKET).
See \fBInteraction with netconfig\fR.
\fBuser_attr\fR\fBgetuserattr\fR(3C)
.TE

.sp
.LP
The following sources can be used:
.sp

.sp
.TS
tab();
cw(2.18i) cw(3.32i) 
lw(2.18i) lw(3.32i) 
.
SourceUses
\fBfiles\fRT{
\fB/etc/hosts\fR, \fB/etc/passwd\fR, \fB/etc/inet/ipnodes\fR, \fB/etc/shadow\fR, \fB/etc/security/auth_attr\fR, \fB/etc/user_attr\fR
T}
\fBnis\fR\fBNIS\fR(\fBYP\fR)
\fBldap\fR\fBLDAP\fR
\fBad\fRActive Directory
\fBdns\fRT{
Valid only for hosts and ipnodes. Uses the Internet Domain Name Service.
T}
\fBmdns\fRT{
Valid only for hosts and ipnodes. Uses the Multicast Domain Name Service.
T}
\fBcompat\fRT{
The \fBcompat\fR source is obsolete. It Implements \fB+\fR and \fB-\fR interaction from SunOS 4. The primary purpose of the \fBcompat\fR pseudo database has been replaced by the \fBpam_list\fR(5) module. See \fBpam_list\fR(5).
T}
.TE

.sp
.LP
\fB/etc/inet/ipnodes\fR is a symbolic link to \fB/etc/hosts\fR.
.sp
.LP
The \fBconfig\fR property group of the \fBsvc:/system/name-service/switch\fR service contains the configuration for the \fBnsswitch.conf\fR file.
.sp
.LP
The \fBconfig/default\fR property sets the default property for all \fBnsswitch\fR databases, while the other properties can be used to override the default property if desired. 
.sp
.LP
For instance:
.sp
.in +2
.nf
config/default set to "files", and
config/host set to "files dns" defaults all
databases to local files database access, except for the
host database which will search files first and dns second,
if dns access is configured.
.fi
.in -2
.sp

.sp
.LP
The following single-valued properties are supported:
.sp
.in +2
.nf
config/default    Default db configuration
config/host       Override for host db
config/password   Override for passwd db
config/group      Override for group db
config/network    Override for network db
config/protocol   Override for protocol db
config/rpc        Override for rpc db
config/ether      Override for ether db
config/netmask    Override for netmask db
config/bootparam  Override for bootparam db
config/publickey  Override for publickey db
config/netgroup   Override for netgroup db
config/automount  Override for automount db
config/alias      Override for alias db
config/service    Override for service db
config/project    Override for project db
config/auth_attr  Override for auth_attr db
config/prof_attr  Override for prof_attr db
config/tnrhtp     Override for tnrhtp db
config/tnrhdb     Override for tnrhdb db
.fi
.in -2
.sp

.sp
.LP
The \fBcompat\fR switch source is considered obsolete. It has been replaced with the \fBpam_list\fR(5) module. The \fBcompat\fR module can still be enabled with the following properties: 
.sp
.in +2
.nf
config/enable_passwd_compat    Enable passwd compat
config/enable_group_compat     Enable group compat
.fi
.in -2
.sp

.sp
.LP
Typically the property values are simple, such as \fB"files" "files nis"\fR. However, when multiple sources are specified, it is sometimes necessary to define precisely the circumstances under which each source is tried. A source can return one of the following codes:
.sp

.sp
.TS
tab();
cw(2.18i) cw(3.32i) 
lw(2.18i) lw(3.32i) 
.
StatusMeaning
\fBSUCCESS\fRRequested database entry was found.
\fBUNAVAIL\fRT{
Source is not configured on this system or internal failure.
T}
\fBNOTFOUND\fRSource responded "\fBno such entry\fR"
\fBTRYAGAIN\fRT{
Source is busy or not responding, might respond to retries.
T}
.TE

.sp
.LP
For each status code, two actions are possible:
.sp

.sp
.TS
tab();
cw(2.18i) cw(3.32i) 
lw(2.18i) lw(3.32i) 
.
ActionMeaning
\fBcontinue\fRTry the next source in the list.
\fBreturn\fRReturn now.
.TE

.sp
.LP
Additionally, for \fBTRYAGAIN\fR only, the following actions are possible:
.sp

.sp
.TS
tab();
cw(2.18i) cw(3.32i) 
lw(2.18i) lw(3.32i) 
.
ActionMeaning
\fBforever\fRRetry the current source forever.
\fIn\fRT{
Retry the current source \fIn\fR more times, where \fIn\fR is an integer between \fB0\fR and \fBMAX_INT\fR (that is, 2.14 billion). After \fIn\fR retries has been exhausted, the \fBTRYAGAIN\fR action transitions to \fBcontinue\fR, until a future request receives a response, at which time \fBTRYAGAIN\fR=\fIn\fR is restored.
T}
.TE

.sp
.LP
The complete syntax of an entry is:
.sp
.in +2
.nf
<entry>     ::= <database> ":" [<source> [<criteria>]]*
<criteria>  ::= "[" <criterion>+ "]"
<criterion> ::= <status> "=" <action>
<status>    ::= "success" | "notfound" | "unavail" | "tryagain"
.fi
.in -2

.sp
.LP
For every status except \fBTRYAGAIN\fR, the action syntax is:
.sp
.in +2
.nf
<action>    ::= "return"  | "continue"
.fi
.in -2

.sp
.LP
For the \fBTRYAGAIN\fR status, the action syntax is:
.sp
.in +2
.nf
<action>    ::= "return"  | "continue" | "forever" | <n>
<n>         ::= 0...MAX_INT          
.fi
.in -2

.sp
.LP
Each property is a single valued string. The <source> names are case-sensitive, but <action> and <status> names are case-insensitive.
.sp
.LP
If an database entry or default is absent, the system defaults to \fB"files"\fR in all situations. Additionally if the name service cache service (svc:/system/name-service-cache) is not enabled, all attemtps access to remote file services can be ignored.
.sp
.LP
The default criteria for \fBDNS\fR and the \fBNIS\fR server in "DNS-forwarding mode" is [\fBSUCCESS\fR=return \fBNOTFOUND\fR=continue \fBUNAVAIL\fR=continue \fBTRYAGAIN\fR=3].
.sp
.LP
The default criteria for all other sources is [\fBSUCCESS\fR=return \fBNOTFOUND\fR=continue \fBUNAVAIL\fR=continue \fBTRYAGAIN\fR=forever].
.sp
.LP
The default, or explicitly specified, criteria are meaningless following the last source in an entry; and they are ignored, since the action is always to return to the caller irrespective of the status code the source returns.
.SS "Interaction with \fBnetconfig\fR"
.sp
.LP
In order to ensure that they all return consistent results, \fBgethostbyname\fR(3NSL), \fBgetaddrinfo\fR(3SOCKET), \fBgetservbyname\fR(3SOCKET), and \fBnetdir_getbyname\fR(3NSL) functions are all implemented in terms of the same internal library function. This function obtains the system-wide source lookup policy for \fBhosts\fR, \fBipnodes\fR, and \fBservices\fR based on the \fBinet\fR family entries in \fBnetconfig\fR(4) and uses the switch entries only if the \fBnetconfig\fR entries have a \fB-\fR (hyphen) in the last column for \fBnametoaddr\fR libraries. See the Notes section in \fBgethostbyname\fR(3NSL) and \fBgetservbyname\fR(3SOCKET) for details.
.SS "Interaction with server in DNS-forwarding Mode"
.sp
.LP
The \fBNIS\fR (\fBYP\fR) server can be run in DNS-forwarding mode, where it forwards lookup requests to \fBDNS\fR for \fBhost-names\fR and \fB-addresses\fR that do not exist in its database. In this case, specifying \fBnis\fR as a source for \fBhosts\fR is sufficient to get \fBDNS\fR lookups; \fBdns\fR need not be specified explicitly as a source.
.SS "Interaction with Password Aging"
.sp
.LP
When password aging is turned on, only a limited set of possible name services are permitted for the \fBpasswd\fR database:
.sp
.ne 2
.mk
.na
\fBpasswd:\fR
.ad
.RS 11n
.rt  
files
.RE

.sp
.ne 2
.mk
.na
\fBpasswd:\fR
.ad
.RS 11n
.rt  
files nis
.RE

.sp
.ne 2
.mk
.na
\fBpasswd:\fR
.ad
.RS 11n
.rt  
files ldap
.RE

.sp
.LP
You can add the \fBad\fR keyword to any of the \fBpasswd\fR configurations listed above. However, you cannot use the \fBpasswd\fR command to change the password of an Active Directory (AD) user. If the \fBad\fR keyword is found in the \fBpasswd\fR entry during a password update operation, it is ignored. To update the password of an AD user, use the \fBkpasswd\fR(1) command.
.sp
.LP
Any other settings causes the \fBpasswd\fR(1) command to fail when it attempts to change the password after expiration and prevents the user from logging in. These are the \fBonly\fR permitted settings when password aging has been turned on. Otherwise, you can work around incorrect \fBpasswd\fR: lines by using the \fB-r repository\fR argument to the \fBpasswd\fR(1) command and using \fBpasswd -r repository\fR to override the \fBnsswitch.conf\fR settings and specify in which name service you want to modify your password.
.SS "Interaction with +/- syntax"
.sp
.LP
The \fB+/-\fR syntax is considered obsolete. It has been replaced both by the \fBnsswitch\fR configuration and the \fBpam_list\fR(5) module. This syntax will be removed in a future release. 
.sp
.LP
Releases prior to SunOS 5.0 did not have the name service switch but did allow the user some policy control. In \fB/etc/passwd\fR one could have entries of the form \fI+user\fR (include the specified user from \fBNIS\fR passwd.byname), \fI-user\fR (exclude the specified user) and \fB+\fR (include everything, except excluded users, from \fBNIS\fR passwd.byname). The desired behavior was often \fBeverything in the file followed by everything in NIS\fR, expressed by a solitary \fB+\fR at the end of \fB/etc/passwd\fR. The switch provides an alternative for this case (\fBpasswd: files nis\fR) that does not require \fB+\fR entries in \fB/etc/passwd\fR and \fB/etc/shadow\fR (the latter is a new addition to SunOS 5.0, see \fBshadow\fR(4)).
.sp
.LP
If this is not sufficient, the \fBNIS/YP\fR compatibility source provides full \fB+/-\fR semantics. It reads \fB/etc/passwd\fR for \fBgetpwnam\fR(3C) functions and \fB/etc/shadow\fR for \fBgetspnam\fR(3C) functions and, if it finds \fB+/-\fR entries, invokes an appropriate source. By default, the source is \fBnis\fR, but this can be overridden by specifying \fBldap\fR as the source for the pseudo-database \fBpasswd_compat\fR.
.sp
.LP
In \fBcompat\fR mode, for every \fB/etc/passwd\fR entry, there must be a corresponding entry in the \fB/etc/shadow\fR file.
.sp
.LP
The NIS/YP compatibility source also provides full \fB+/-\fR semantics for \fBgroup\fR; the relevant pseudo-database is \fBgroup_compat\fR.
.SS "Hard-wired Policies"
.sp
.LP
The compiled-in default entries for all databases is \fB"files"\fR.
.SS "Useful Configuration Notes"
.sp
.LP
The files source for the ipnodes and hosts databases is identical, as \fB/etc/inet/ipnodes\fR is a symbolic link to \fB/etc/hosts\fR. The host property is used for host lookups. 
.sp
.LP
When using Active Directory, \fBdns\fR is required to perform hosts resolution.
.sp
.LP
In order to get information from the Internet Domain Name Service for hosts that are not listed in the enterprise level name service \fBLDAP\fR, use the following configuration and set up the \fB/etc/resolv.conf\fR file (see \fBresolv.conf\fR(4) for more details):
.sp
.ne 2
.mk
.na
\fBhosts:\fR
.ad
.RS 10n
.rt  
files dns
.RE

.SS "Enumeration - \fBgetXXXent()\fR"
.sp
.LP
Many of the databases have enumeration functions: \fBpasswd\fR has \fBgetpwent()\fR, \fBhosts\fR has \fBgethostent()\fR, and so on. These were reasonable when the only source was \fBfiles\fR but often make little sense for hierarchically structured sources that contain large numbers of entries, much less for multiple sources. The interfaces are still provided and the implementations strive to provide reasonable results, but the data returned can be incomplete (enumeration for \fBhosts\fR is simply not supported by the \fBdns\fR source), inconsistent (if multiple sources are used), formatted in an unexpected fashion (for a host with a canonical name and three aliases, a source might return four hostents, and they might not be consecutive), or very expensive (enumerating a \fBpasswd\fR database of 5,000 users is probably a bad idea). Furthermore, multiple threads in the same process using the same reentrant enumeration function (\fBgetXXXent_r()\fR are supported beginning with SunOS 5.3) share the same enumeration position; if they interleave calls, they enumerate disjoint subsets of the same database.
.sp
.LP
In general, the use of the enumeration functions is deprecated. In the case of \fBpasswd\fR, \fBshadow\fR, and \fBgroup\fR, it might sometimes be appropriate to use \fBfgetgrent()\fR, \fBfgetpwent()\fR, and \fBfgetspent()\fR (see \fBgetgrnam\fR(3C), \fBgetpwnam\fR(3C), and \fBgetspnam\fR(3C), respectively), which use only the \fBfiles\fR source.
.SH FILES
.sp
.LP
A source named SSS is implemented by a shared object named \fBnss_SSS.so.1\fR that resides in \fB/usr/lib\fR.
.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.conf\fR\fR
.ad
.RS 28n
.rt  
Configuration file. (Obsolete.)
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_compat.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBcompat\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_dns.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBdns\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_files.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBfiles\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_mdns.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBmdns\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_nis.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBnis\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_ldap.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBldap\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_ad.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements ad source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/usr/lib/nss_user.so.1\fR\fR
.ad
.RS 28n
.rt  
Implements \fBuser\fR source.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/netconfig\fR\fR
.ad
.RS 28n
.rt  
Configuration file for \fBnetdir\fR(3NSL) functions that redirects hosts/devices policy to the switch.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.files\fR\fR
.ad
.RS 28n
.rt  
Sample configuration file that uses \fBfiles\fR only.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.nis\fR\fR
.ad
.RS 28n
.rt  
Sample configuration file that uses \fBfiles\fR and \fBnis\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.ldap\fR\fR
.ad
.RS 28n
.rt  
Sample configuration file that uses \fBfiles\fR and \fBldap\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.ad\fR\fR
.ad
.RS 28n
.rt  
Sample configuration file that uses \fBfiles\fR and \fBad\fR.
.RE

.sp
.ne 2
.mk
.na
\fB\fB/etc/nsswitch.dns\fR\fR
.ad
.RS 28n
.rt  
Sample configuration file that uses \fBfiles\fR, \fBdns\fR and \fBmdns\fR (\fBdns\fR and \fBmdns\fR only for hosts).
.RE

.SH SEE ALSO
.sp
.LP
\fBkpasswd\fR(1), \fBnewtask\fR(1), \fBpasswd\fR(1), \fBautomount\fR(1M), \fBifconfig\fR(1M), \fBmdnsd\fR(1M), \fBrpc.bootparamd\fR(1M), \fBsendmail\fR(1M), \fBgetgrnam\fR(3C), \fBgetnetgrent\fR(3C), \fBgetpwnam\fR(3C), \fBgetspnam\fR(3C), \fBgethostbyname\fR(3NSL), \fBgetpublickey\fR(3NSL), \fBgetrpcbyname\fR(3NSL), \fBnetdir\fR(3NSL), \fBsecure_rpc\fR(3NSL), \fBgetprojent\fR(3PROJECT), \fBgetdefaultproj\fR(3PROJECT), \fBinproj\fR(3PROJECT), \fBsetproject\fR(3PROJECT), \fBgetauthnam\fR(3C), \fBgetexecprof\fR(3C), \fBgetprofnam\fR(3C), \fBgetuserattr\fR(3C), \fBgetusernam\fR(3C), \fBethers\fR(3SOCKET), \fBgetaddrinfo\fR(3SOCKET), \fBgetnetbyname\fR(3SOCKET), \fBgetprotobyname\fR(3SOCKET), \fBgetservbyname\fR(3SOCKET), \fBauth_attr\fR(4), \fBhosts\fR(4), \fBnetconfig\fR(4), \fBproject\fR(4), \fBresolv.conf\fR(4), \fBuser_attr\fR(4), \fBypfiles\fR(4), \fBad\fR(5), \fBldap\fR(5), \fBpam_list\fR(5)
.SH NOTES
.sp
.LP
Within each process that uses \fBnsswitch.conf\fR, the entire file is read only once; if the file is later changed, the process continues using the old configuration.
.sp
.LP
The use of both \fBnis\fR and \fBldap\fR as sources for the same database is strongly discouraged since both the name services are expected to store similar information and the lookups on the database can yield different results depending on which name service is operational at the time of the request.
.sp
.LP
Do not use the \fBldap\fR and \fBad\fR keywords together when the Solaris LDAP client uses schema mapping to talk to Active Directory. 
.sp
.LP
Misspelled names of sources and databases are treated as legitimate names of (most likely nonexistent) sources and databases.
.sp
.LP
\fBnsswitch.conf\fR does not control the name service configuration for everything in Solaris. 
.sp
.LP
The following functions do \fBnot\fR use the switch: \fBfgetgrent\fR(3C), \fBfgetprojent\fR(3PROJECT), \fBfgetpwent\fR(3C), \fBfgetspent\fR(3C), \fBgetpw\fR(3C), \fBputpwent\fR(3C), \fBshadow\fR(4).


usr/src/lib/nsswitch/pynss/common/nsswitch.py
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License (the "License").
# You may not use this file except in compliance with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#

#
# Copyright (c) 2011, Oracle and/or its affiliates. All rights reserved.
#

"""Name Service Switch classes used by nscfg."""

from __future__ import print_function
from .nssbase import Nssbase

class Nsswitch(Nssbase):
    """Name Service Switch class. Represents the configuration
       of the name service switch.  See nsswitch.conf(4) for details."""

    SERVICE = 'svc:/system/name-service/switch'
    LEGACY = 'nsswitch.conf'
    LEGACYDIR = '/etc'
    DEFPG = 'config'
    DEFPROP = 'default'
    COPY_YR = '1991'			# Initial Copyright year
    DOC = "# See nsswitch.conf(4) for details.\n"

	# This class only supports the nsswitch_config property group
	# The Nssfiles class supports the files_config property group
    ALLPGS = [ 'config' ]
    ALLPROPS = {'default': 'astring',
	'host': 'astring',
	'password': 'astring',
	'group': 'astring',
	'network': 'astring',
	'protocol': 'astring',
	'rpc': 'astring',
	'ether': 'astring',
	'netmask': 'astring',
	'bootparam': 'astring',
	'publickey': 'astring',
	'netgroup': 'astring',
	'automount': 'astring',
	'alias': 'astring',
	'service': 'astring',
	'printer': 'astring',
	'project': 'astring',
	'auth_attr': 'astring',
	'prof_attr': 'astring',
	'tnrhtp': 'astring',
	'tnrhdb': 'astring',
	'enable_passwd_compat': 'boolean',
	'enable_group_compat': 'boolean', }

    # nsswitch.conf order is not important, but administrators
    # historically tend to expect it in this specific order.

    DBORDER = ('passwd', 'group', 'hosts', 'ipnodes', 'networks',
	     'protocols', 'rpc', 'ethers', 'netmasks', 'bootparams',
	     'publickey', 'netgroup', 'automount', 'aliases',
	     'services', 'printers', 'project', 'auth_attr',
	     'prof_attr', 'tnrhtp', 'tnrhdb',)

    DEFPVAL = 'files'
    DEFPTYPE = 'astring'

    PWD_COMPAT = 'passwd_compat'
    EN_PWD_COMPAT = 'enable_passwd_compat'
    GRP_COMPAT = 'group_compat'
    EN_GRP_COMPAT = 'enable_group_compat'

    UNCPROP = (
	    ( 'default', 'astring', 'files' ),
	    ( 'printer', 'astring', 'user files' ),
	)

    VALUE_AUTH = 'solaris.smf.value.name-service.switch'

    def __init__(self):
	Nssbase.__init__(self)

    def export_from_smf(self):
	"""Export from SMF.  Re-generate nsswitch.conf file"""
	self.print_msg('exporting nsswitch.conf legacy...')
	# Map SMF properties to DB
	db = {}
	pwd_config = False
	pval = self.get_prop_val(prop='enable_passwd_compat')
	if type(pval) == type('') and pval == 'true':
	    pwd_config = True
	grp_config = False
	pval = self.get_prop_val(prop='enable_group_compat')
	if type(pval) == type('') and pval == 'true':
	    grp_config = True
	dval = self.get_prop_val()		# Get default pg/prop
	if dval == None or dval == '':
	    dval = 'files'			# fallback to files only
	self.traceit('dval ', dval)
	for k in self.DBMAP.keys():
	    self.traceit('k= ', k)
	    db[k] = dval			# Force default
	    pval = self.get_prop_val(prop=self.DBMAP[k])
	    self.traceit('  pval[k]= ', pval)
	    if pval != None and pval != '':
		db[k] = pval			# Set Specific
	    self.traceit('db[k]= ', db[k])
	data = ''
	for k in self.DBORDER:			# Could use db.keys()
	    if k == 'passwd' and pwd_config:
		data += "%s:\t%s\n" % (k, 'compat')
		data += "%s:\t%s\n" % (self.PWD_COMPAT, db[k])
	    elif k == 'group' and grp_config:
		data += "%s:\t%s\n" % (k, 'compat')
		data += "%s:\t%s\n" % (self.GRP_COMPAT, db[k])
	    else:
		data += "%s:\t%s\n" % (k, db[k])
	self.traceit('data= ', data)
	# Save to temp file, with CDDL header
        if not self.save_to_tmp(True, data):
	    self.err_msg('FAIL save_to_tmp')
	    return self.FAIL
	ret = self.tmp_to_legacy()
	if ret == self.FAIL:
	    self.print_msg('fail move to legacy...')
	    self.err_msg('FAIL tmp_to_legacy')
	    self.unlink_tmp()
	    return self.FAIL
	self.print_msg('successful export.')
	return ret

    def import_to_smf(self):
	"""Import to SMF.  Configure SMF from nsswitch.conf file.
	   The default props are always reset."""
	self.print_msg('importing legacy nsswitch.conf...')
	db = []
	pwd_config = False
	pwd_val = None
	grp_config = False
	grp_val = None
	lines = self.load_legacy()
	if lines == None:			# defaults always exist
	    lines = []
	self.traceit('    loaded legacy: ', lines)
	for l in lines:
	    if l.startswith('#'):		# Skip comment lines
		continue
	    idx = l.find('#')
	    if idx > 0:
		l = l[:idx]
	    while l.endswith('\t') or l.endswith(' '):
		l = l[:-1]
	    if l == '':				# Skip enpty lines
		continue
	    self.traceit('    PROCESS LINE: ', l)
	    vals = l.split(None, 1)
	    if len(vals) != 2:
		continue
	    try:
		self.traceit('    trying: ', vals)
		key = vals[0]			# verify key and values
		if not key.endswith(':'):
		    continue			# Not a valid DB key
		key = key[:-1]
		if key == 'ipnodes':
		    continue			# use 'hosts' not 'ipnodes'
		if key == 'passwd' and vals[1] == 'compat':
		    pwd_config = True
		    db.append((self.EN_PWD_COMPAT, \
			      self.ALLPROPS[self.EN_PWD_COMPAT], 'true'))
		    continue
		if key == 'group' and vals[1] == 'compat':
		    grp_config = True
		    db.append((self.EN_GRP_COMPAT, \
			      self.ALLPROPS[self.EN_GRP_COMPAT], 'true'))
		    continue
		if key == self.PWD_COMPAT:
		    if pwd_config == False:
			continue		# ignore (actually error)
		    map = self.DBMAP['passwd']
		elif key == self.GRP_COMPAT:
		    if grp_config == False:
			continue		# ignore (actually error)
		    map = self.DBMAP['group']
		else:
		    map = self.DBMAP[key]
		if map == 'password':
		    pwd_val = vals[1]
		if map == 'group':
		    grp_val = vals[1]
		db.append((map, self.ALLPROPS[map], vals[1]))
	    except:
		continue
	self.traceit('    processed db: ', db)
	# Backwards compatibility checks
	# Check and adjust for passwd: compat w/ no passwd_compat line.
	if pwd_config and pwd_val == None:
	    map = 'password'
	    db.append((map, self.ALLPROPS[map], "nis"))
	if grp_config and grp_val == None:
	    map = 'group'
	    db.append((map, self.ALLPROPS[map], "nis"))
	# db now holds all key/value pairs to send to smf
	# Exit before write processing
	if self.no_write:
	    return self.SUCCESS
	self.print_msg('    delete customizations.')
	if not self.delcust_pg(self.DEFPG):
	    return self.FAIL
	# Commit the property group
	if not self.commit():
	    return self.FAIL
	self.print_msg('    loading pg...')
	err = False
	for (prop, ptype, val) in db:
	    self.traceit('      adding: ', prop, ptype, val)
	    if val == self.DEFPVAL:
		continue			# Skip over default 'files'
	    if not self.add_prop_val(self.DEFPG, prop, ptype, val):
		self.traceit('      ERR adding: ', prop, ptype, val)
		return self.FAIL
	# Commit the property group
	self.print_msg('    committing pg...')
	if not self.commit(self.DEFPG):
	    return self.FAIL
	# validate
	self.print_msg('    validating pg...')
	if not self.validate():
	    return self.FAIL
	self.print_msg('successful import.')
	return self.SUCCESS

    def unconfig_smf(self):
	"""Unconfigure SMF.  Reset DEFPG."""
	self.print_msg('unconfiguring nsswitch...')
	self.print_msg('Delete customizations.')
	if not self.delcust_pg(self.DEFPG):
	    return self.FAIL
	# Commit the property group
	if not self.commit():
	    return self.FAIL
	self.print_msg('successful unconfigure.')
	return self.SUCCESS

